---
title: "custom_kraken2_output_visual"
author: "Rachel Xu"
date: "6/22/2020"
output: html_document
---
### PATH to custom output result:
```{r}
path = "/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/"
setwd(path)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = path)
```

```{r}
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyverse)
library(plotly)
library("phyloseq")
```
###include a color palette for ggplot
```{r}
manual_color <-c("#4361EE","#FFADAD","#FDFFB6","#FFD6A5","#CAFFBF","#9BF6FF","#A0C4FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C","#F94144","#80B918","#FEE440")

```
## Kraken2 kreport:
    **1) Absolute Reads**
    **2) root: celluar organism (Eukaryota + Bacteria) + Viruses**
    **3) all_reads: Eukaryota + Bacteria + Viruses + unclassified**
    
### Domain Classification 
```{r}

cus_path = paste(path,"custom/absolute/",sep="")
files <- list.files(cus_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(cus_path,file,sep=""),sep="\t",header=FALSE)
  # add "| trimws(V6) == "other sequences"" to subset if counting total number of reads
  domain_absolute <- current_sample %>% subset(V4 == "D" | V4 == "U" | trimws(V6) == "other sequences") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))

  colnames(domain_absolute) <- c("Domain", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    domain_all_samples <-  domain_absolute

  } else {
    domain_all_samples <- full_join(domain_all_samples, domain_absolute,by="Domain")

  }
  i = i + 1
}

# save(domain_all_samples, file ="custom_absolute_domain.Rdata")

domain_all_samples[is.na(domain_all_samples)] <- 0
keys_to_gather <- colnames(domain_all_samples)[-1] # gather samples, exclude "Domain" from colnames

# prepare dataframe for ggplot
domain_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")


domain_all_samples$samples <- factor(domain_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
domain_all_samples$Domain <- factor(domain_all_samples$Domain, level=c("unclassified","Eukaryota","Bacteria","Viruses","Archaea","other sequences"))

# Total number of reads for each sample (unclassified + three domains + other sequences)
domain_all_samples %>% group_by(samples) %>% summarise(total_reads=sum(Reads)) -> cus_total_reads
as.data.frame(list(mean_reads=mean(cus_total_reads$total_reads),std=sd(cus_total_reads$total_reads))) -> cus_total_reads_summary # avg number of reads across sample

# Total number of reads for each sample (unclassified + three domains/ exclude "other sequences")
domain_all_samples %>% subset(Domain != "other sequences") %>% group_by(samples) %>% summarise(total_reads=sum(Reads)) -> cus_total_reads_no_other
as.data.frame(list(mean_reads=mean(cus_total_reads_no_other$total_reads),std=sd(cus_total_reads_no_other$total_reads))) -> cus_total_reads_no_other_summary # avg number of reads across sample
cus_total_reads_no_other
cus_total_reads_no_other_summary

# other sequences summary
domain_all_samples %>% subset(Domain == "other sequences") -> cus_other_seq
cus_other_seq %>% summarise(mean_other=mean(Reads),std=sd(Reads)) -> other_seq_cus_summary

# unclassified reads summary
domain_all_samples %>% subset(Domain == "unclassified") -> cus_unc
cus_unc %>%summarise(mean_unc=mean(Reads),std=sd(Reads)) -> cus_unc_summary

cus_total_reads_summary # avg number of reads across sample
other_seq_cus_summary
cus_unc_summary
cus_total_reads
```

```{r}
plot <- ggplot(domain_all_samples, aes(x=samples,y=Reads,fill=Domain)) +
  geom_bar(stat="identity") + 
  theme_classic() +
  theme(legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank(), axis.text.x= element_blank(),legend.position="none") +
  scale_fill_manual(values=manual_color)

plot
ggplotly(plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_cus_domain.png", width = 8, height=6)

#### plot relative abundance 
domain_all_samples %>%
  group_by(samples) %>%
  mutate(Total=sum(Reads)) %>%
  mutate(Relative=Reads/Total) -> domain_relative


relative_plot <- ggplot(domain_relative, aes(x=samples,y=Relative,fill=Domain)) +
  geom_bar(stat="identity") + 
  theme_classic() +
  theme(legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank(), axis.text.x= element_blank(),legend.position="none") +
  scale_y_continuous(labels = scales::percent_format()) +
  scale_fill_manual(values=manual_color)
ggplotly(relative_plot)
relative_plot
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_cus_domain_rel.png", width = 8, height=6)

domain_relative
domain_relative %>% group_by(Domain) %>% summarise(mean_reads = mean(Reads), std_reads = sd(Reads), mean_Percentage= mean(Relative), std_Percentage=sd(Relative))
```

### Classify phylum, excluding unclassified and Eukaryote reads to increase local resolution
    
```{r}
absolute_path = paste(path,"custom/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  phylum_absolute <- current_sample %>% subset(V4 == "P") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  phylum_absolute <- phylum_absolute %>% subset(c(trimed!="Chordata")) #%>% mutate( V2 >= 100) # deselect phylum belongs to Eukrayota
  colnames(phylum_absolute) <- c("Phylum", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    phylum_all_samples <-  phylum_absolute

  } else {
    phylum_all_samples <- full_join(phylum_all_samples, phylum_absolute,by="Phylum")

  }
  i = i + 1
}

# save(phylum_all_samples, file ="standard_absolute_phylum_micro_only.Rdata")

phylum_all_samples[is.na(phylum_all_samples)] <- 0
keys_to_gather <- colnames(phylum_all_samples)[-1] # gather samples, exclude "Phylum" from colnames

# prepare dataframe for ggplot
phylum_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

all_phylums <- unique(phylum_all_samples$Phylum) # all phylums classified
# phylums I want to factored into order so they can be assigned with a specific color during plotting
order_phylum <- c("Proteobacteria","Actinobacteria","Cyanobacteria","Firmicutes","Bacteroidetes","Uroviricota","Tenericutes","Spirochaetes","Fusobacteria","Chlamydiae","Aquificae","Chloroflexi","Thermotogae","Artverviricota","Peploviricota","Deinococcus-Thermus","Planctomycetes","Euryarchaeota","Gemmatimonadetes","Crenarchaeota") 
remaining_phylum <- all_phylums %>% data.frame() %>% subset(!c(all_phylums %in% order_phylum)) 
remaining_phylum <- remaining_phylum$.

phylum_all_samples$samples <- factor(phylum_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
phylum_all_samples$Phylum <- factor(phylum_all_samples$Phylum, level=c(order_phylum,remaining_phylum))


# subsetting taxonomy with more than 1% in relative composition from each sample
phylum_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>% subset(Percentage >= 0.001) ->major_phylum
major_phylum


```
# plot phylum
```{r}


plot <- ggplot(major_phylum, aes(x=samples,y=Reads,fill=Phylum)) +
  geom_bar(stat="identity") + 
  labs(title = "Kraken2 Microbiome Reads Profile with Customized DB (Phylum)", y="Number of Reads", x = "Samples") + 
   theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm"))+
  scale_fill_manual(values=manual_color)
ggplotly(plot)
plot

ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_cus_phylum.png",width=8, height=6)




relative_plot <- ggplot(major_phylum, aes(x=samples,y=Percentage,fill=Phylum)) +
  geom_bar(stat="identity") + 
  labs(title = "Kraken2 Microbiome Composition Profile with Customized DB (Phylum)", y="Percentage", x = "Samples") + 
 theme(axis.text.x = element_text(angle=90),legend.text = element_text(size = 8),legend.key.size = unit(0.4,"cm")) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x=element_text(angle=90)) + 
  scale_fill_manual(values=manual_color)
ggplotly(relative_plot)
relative_plot
# ggplotly(relative_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_cus_phylum_rel.png",width=8, height=6)

```

### Classify genus, excluding unclassified and Eukaryote reads to increase local resolution

```{r}
absolute_path = paste(path,"custom/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  genus_absolute <- current_sample %>% subset(V4 == "G") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  genus_absolute <- genus_absolute %>% subset(trimed!="Rattus" & trimed !="Homo") #%>% subset(V2 >= 300) # deselect genus belongs to Eukrayota
  colnames(genus_absolute) <- c("genus", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    genus_all_samples <-  genus_absolute

  } else {
    genus_all_samples <- full_join(genus_all_samples, genus_absolute,by="genus")

  }
  i = i + 1
}

# save(genus_all_samples, file ="standard_absolute_genus_micro_only.Rdata")

genus_all_samples[is.na(genus_all_samples)] <- 0
keys_to_gather <- colnames(genus_all_samples)[-1] # gather samples, exclude "genus" from colnames

# prepare dataframe for ggplot
genus_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

# genus I want to factored in order so they can be assigned with a specific color during plotting
all_genus <- unique(genus_all_samples$genus) # allgenus classified
order_genus <- c("Xanthomonas","Pseudomonas","Pasteurella","Streptomyces","Calothrix","Andhravirus","Mycoplasma","Spiroplasma","Bacillus","Staphylococcus","Lactobacillus","Clostridium","Mycobacterium","Bordetella","Leptospira","Yersinia","Escherichia","Bartonella","Bacteroides","Microcystis","Nostoc","Mycobacteroides","Alistipes","Parabacteroides","Anabaena","Dolichospermum","Faecalibacterium","Sphaerospermopsis","Prevotella","Achromobacter") 
remaining_genus <- all_genus %>% data.frame() %>% subset(!c(all_genus %in% order_genus)) # deselect genus subsetted to put in order
remaining_genus <- remaining_genus$. # data.frame to vector

genus_all_samples$samples <- factor(genus_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
genus_all_samples$genus <- factor(genus_all_samples$genus, level=c(order_genus,remaining_genus))

# subsetting taxonomy with more than 1% in relative composition from each sample 
genus_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>% subset(Percentage >= 0.010) ->major_genus

# number of reads classified into microbiome composition (exclude homo and rattus)
genus_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>%subset(genus != "Rattus" & genus !="Homo") %>% summarise(sum_reads=sum(Reads), sum_percentage=sum(Percentage)) -> cus_micro_comp
cus_micro_comp
```

```{r}
manual_color <-c("#54478C","#97CBF0ff","#F94144","#80B918","#FEE440","#4361EE","#FFADAD","#FFD6A5","#FDFFB6","#CAFFBF","#9BF6FF","#A0C4FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#41EAD4","#FBFF12","#54478C")

plot <- ggplot(major_genus, aes(x=samples,y=Reads,fill=genus)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Reads Classification with Customized Kraken2 DB (Genus)", y="Number of Reads", x = "Samples") + 
  theme(legend.text = element_text(size =8 ),legend.key.size =unit(0.4,"cm"))+
  scale_fill_manual(values=manual_color)
ggplotly(plot)
plot
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Absolute/kraken2_cus_genus.png",width=10, height=6)



relative_plot <- ggplot(major_genus, aes(x=samples,y=Percentage,fill=genus)) +
  geom_bar(stat="identity") + 
  labs(title = "Metagenomic Relative Abundance with Customized Kraken2 DB (Genus)", y="Percentage", x = "Samples") + 
  theme(legend.text = element_text(size =8 ),legend.key.size =unit(0.4,"cm")) +
  scale_y_continuous(labels = scales::percent_format()) +
  theme(axis.text.x=element_text(angle=90)) + 
  scale_fill_manual(values=manual_color)
relative_plot
ggplotly(relative_plot)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/Plots/Relative/kraken2_cus_genus_rel.png",width=10, height=6)


```

## Phyloseq Package for Downstream Analysis

### Data Wrangling to format for phyloseq analysis
```{r}
path = "/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/"
absolute_path = paste(path,"custom/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  bacteria_profile <- current_sample %>% mutate(trimed = trimws(V6)) %>% subset(V4 == "G") %>% select(c(trimed,V2))  %>% subset(! trimed %in% "Rattus" &  ! trimed %in% "Homo")# only keep rank that has certain taxonomy
  rownames(bacteria_profile) <- NULL # reorder rownames after subsetting
  colnames(bacteria_profile) <- c("OTU",unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    all_samples_w_taxa <-  bacteria_profile

  } else {
    all_samples_w_taxa <- full_join(all_samples_w_taxa, bacteria_profile,by=c("OTU"))

  }
  i = i + 1
}



# make the OTU names the rownames
rownames(all_samples_w_taxa) <- all_samples_w_taxa$OTU
# write.csv(all_samples_w_taxa$OTU,paste0(path, "kraken2_genus_tax.csv"), quote=FALSE, row.names = FALSE, col.names = NA)
all_samples_w_taxa[is.na(all_samples_w_taxa)]<-0

```

### input into phyloseq objects
```{r}

otu_tab <- all_samples_w_taxa[,-1]
OTU <- otu_table(otu_tab,taxa_are_rows = T) # OTU table, abundance level for each taxa for each sample

# to obtain a tax table,I retrieved the lineage for all genus in bracken results
# use this tool : https://www.ncbi.nlm.nih.gov/Taxonomy/TaxIdentifier/tax_identifier.cgi
# since we only need Phylum and kingdom, next time figure out a way to query phylum and kingdom of each genus instead of query the entire lineage
tax <- read.csv("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/kraken_genus_lineage.csv", header = FALSE)
rownames(tax) <- rownames(otu_tab)
colnames(tax) <- c("Genus","Phylum","Domain")
TAX <- tax_table(as.matrix(tax)) # TAX table, taxonomy lineage for each taxa from the OTU table

# add sample's metadata
meta_sample <- read.csv("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/kraken2_sample_meta.csv")
row.names(meta_sample) <- meta_sample$X
meta_sample <- meta_sample[,-1]
sample_meta <- sample_data(meta_sample)



# merge all information together
cus_kraken2_physeq <- phyloseq(OTU,TAX,sample_meta)

library("ape")
random_tree_kraken2 = rtree(ntaxa(cus_kraken2_physeq), rooted=TRUE, tip.label=taxa_names(cus_kraken2_physeq))
# plot_bar(cus_kraken2_physeq, fill="Domain")
# plot_heatmap(cus_kraken2_physeq, taxa.label="Domain")
# plot_tree(cus_kraken2_physeq, color="Domain", shape="Tissue", size="abundance")

# merge all information together
cus_kraken2_physeq <- phyloseq(OTU,TAX,sample_meta,random_tree_kraken2)
```

### Alpha Diversity (Within Sample diversity)
- tutorial: https://grunwaldlab.github.io/analysis_of_microbiome_community_data_in_r/07--diversity_stats.html

- **Shannon**: How difficult it is to predict the identity of a randomly chosen individual.
- **Simpson**: The probability that two randomly chosen individuals are the same species.
- **Inverse Simpson**: This is a bit confusing to think about. Assuming a theoretically community where all species were equally abundant, this would be the number of species needed to have the same Simpson index value for the community being analyzed.

```{r}

library(vegan)
# margin = 2, means samples are the col names
alpha_diversity <- data.frame(Shannon = diversity(otu_tab, MARGIN =2, index="shannon"), Simpson = diversity(otu_tab, MARGIN =2, index="simpson"), Inverse_Simpson = diversity(otu_tab, MARGIN =2, index="invsimpson")) 

# plot Alpha diversity 
# plot Alpha diversity 
plot_richness(cus_kraken2_physeq, x= "Tissue", color="Subject",measures = c("Shannon", "Simpson", "InvSimpson"),title = "Alpha diversity for each sample")

cus_kraken2_physeq
```

###Beta Diversity (between samples)
https://grunwaldlab.github.io/analysis_of_microbiome_community_data_in_r/07--diversity_stats.html
- Bray–Curtis: The sum of lesser counts for species present in both communities divided by the sum of all counts in both communities. This can be thought of as a quantitative version of the Sørensen index.
- Weighted Unifrac: The fraction of the phylogenetic tree branch lengths shared by the two communities, weighted by the counts of organisms, so more abundant organisms have a greater influence.
```{r}

# no margin option like alpha diversity, use transpose for samples as colnames
# bray-curtis
beta_dist <-vegdist(t(otu_tab), index="bray") # pairwise comparison between samples

# plot beta diversity using ordination, a way to display high dimensional data
# capture the information in many dimensions by in a smaller number of "artifical" dimensions
mds <- metaMDS(beta_dist)
mds_kraken2 <- as.data.frame(mds$points)
mds_kraken2$samples <- rownames(mds_kraken2)
meta_sample$samples <- rownames(meta_sample)
mds_kraken2_meta <- left_join(meta_sample,mds_kraken2, by="samples")

ggplot(mds_kraken2_meta,aes(x=MDS1,y=MDS2,color=Subject))+
  geom_point()
```


### Differential Abundant with Deseq2
- load DESeq2 package
```{r}
library("DESeq2")

```

- differentially abundant analysis with deseq2
```{r}
desseq_kraken2 <- phyloseq_to_deseq2(cus_kraken2_physeq, ~ Tissue)
desseq_kraken2
kraken_deseq <- DESeq(desseq_kraken2, test="Wald", fitType="parametric")

# pairwise comparison between 2 tissues
Kidney_lung <- results(kraken_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Lung"))
kidney_spleen <- results(kraken_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Spleen"))
lung_spleen <- results(kraken_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Lung","Spleen"))

alpha = 0.01 # significance threhold

# change pairwise comparison results for comparing two different tissues
sigtax_kraken2 <- Kidney_lung[which(Kidney_lung$padj < alpha),] # get tax below the significance thresholf
sigtax_kraken2$Genus <- rownames(sigtax_kraken2) # add a column so genus tax can combine with upper lineages

# add lineages to significant taxa
kraken2_deseq_lineage <- left_join(as.data.frame(sigtax_kraken2), tax)
dim(kraken2_deseq_lineage)
unique(kraken2_deseq_lineage$Phylum)
meta_sample

```

- Plot significantly abundant taxa from the samples
```{r}
# find the maximum log2fold change from deseq2 output for each cell (tapply)
x<-tapply(kraken2_deseq_lineage$log2FoldChange,kraken2_deseq_lineage$Phylum,function(x)max(x))
x<- sort(x,TRUE) # sort these identified phyla in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
kraken2_deseq_lineage$Phylum <- factor(as.character(kraken2_deseq_lineage$Phylum),levels = names(x))

# perform the same ordering procedure with genus too
x<-tapply(kraken2_deseq_lineage$log2FoldChange,kraken2_deseq_lineage$Genus,function(x)max(x))
x<- sort(x,TRUE) # sort these identified genus in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
kraken2_deseq_lineage$Genus <- factor(as.character(kraken2_deseq_lineage$Genus),levels = names(x))

ggplot(kraken2_deseq_lineage, aes(x=Genus, y=log2FoldChange, color= Phylum))+ geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
```