---
title: "Compare_statistical_analysis"
output: html_document
---
# Packages
```{r}
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyverse)
library(plotly)
library("phyloseq")
library(ggrepel)
```

```{r}


sig_phylum_color <-
  c("#F8766D","#D89000", "#85AD00" ,"#00A5FF","#DC71FA","#00C0B8","#39B600")
            
```
# Kraken2 
### Data Wrangling to format for phyloseq analysis
```{r}
path = "/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/"
absolute_path = paste(path,"custom/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  bacteria_profile <- current_sample %>% mutate(trimed = trimws(V6)) %>% subset(V4 == "G") %>% select(c(trimed,V2))  %>% subset(! trimed %in% "Rattus" &  ! trimed %in% "Homo")# only keep rank that has certain taxonomy
  rownames(bacteria_profile) <- NULL # reorder rownames after subsetting
  colnames(bacteria_profile) <- c("OTU",unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    all_samples_w_taxa <-  bacteria_profile

  } else {
    all_samples_w_taxa <- full_join(all_samples_w_taxa, bacteria_profile,by=c("OTU"))

  }
  i = i + 1
}



# make the OTU names the rownames
rownames(all_samples_w_taxa) <- all_samples_w_taxa$OTU
# write.csv(all_samples_w_taxa$OTU,paste0(path, "kraken2_genus_tax.csv"), quote=FALSE, row.names = FALSE, col.names = NA)
all_samples_w_taxa[is.na(all_samples_w_taxa)]<-0

```

### input into phyloseq objects
```{r}

otu_tab <- all_samples_w_taxa[,-1]
OTU <- otu_table(otu_tab,taxa_are_rows = T) # OTU table, abundance level for each taxa for each sample

# to obtain a tax table,I retrieved the lineage for all genus in bracken results
# use this tool : https://www.ncbi.nlm.nih.gov/Taxonomy/TaxIdentifier/tax_identifier.cgi
# since we only need Phylum and kingdom, next time figure out a way to query phylum and kingdom of each genus instead of query the entire lineage
tax <- read.csv("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/kraken_genus_lineage.csv", header = FALSE)
rownames(tax) <- rownames(otu_tab)
colnames(tax) <- c("Genus","Phylum","Domain")
TAX <- tax_table(as.matrix(tax)) # TAX table, taxonomy lineage for each taxa from the OTU table

# add sample's metadata
meta_sample <- read.csv("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/kraken2_sample_meta.csv")
row.names(meta_sample) <- meta_sample$X
meta_sample <- meta_sample[,-1]
sample_meta <- sample_data(meta_sample)



# merge all information together
cus_kraken2_physeq <- phyloseq(OTU,TAX,sample_meta)

library("ape")
random_tree_kraken2 = rtree(ntaxa(cus_kraken2_physeq), rooted=TRUE, tip.label=taxa_names(cus_kraken2_physeq))
# plot_bar(cus_kraken2_physeq, fill="Domain")
# plot_heatmap(cus_kraken2_physeq, taxa.label="Domain")
# plot_tree(cus_kraken2_physeq, color="Domain", shape="Tissue", size="abundance")

# merge all information together
cus_kraken2_physeq <- phyloseq(OTU,TAX,sample_meta,random_tree_kraken2)
```

### Alpha Diversity (Within Sample diversity)
- tutorial: https://grunwaldlab.github.io/analysis_of_microbiome_community_data_in_r/07--diversity_stats.html

- **Shannon**: How difficult it is to predict the identity of a randomly chosen individual.
- **Simpson**: The probability that two randomly chosen individuals are the same species.
- **Inverse Simpson**: This is a bit confusing to think about. Assuming a theoretically community where all species were equally abundant, this would be the number of species needed to have the same Simpson index value for the community being analyzed.

```{r}

library(vegan)
# margin = 2, means samples are the col names
alpha_diversity_kraken2 <- data.frame(Shannon = diversity(otu_tab, MARGIN =2, index="shannon"), Simpson = diversity(otu_tab, MARGIN =2, index="simpson"), Inverse_Simpson = diversity(otu_tab, MARGIN =2, index="invsimpson")) 

# plot Alpha diversity 
# plot Alpha diversity 
plot_richness(cus_kraken2_physeq, x= "Tissue", color="Subject",measures = c("Shannon", "Simpson", "InvSimpson"))+
  geom_point(size=5) +
   theme_classic()+
  theme(legend.key.size = unit(1.0, "cm"), legend.text = element_text(size=13),legend.title = element_text(size=14,face="bold"), strip.text = element_text(face="bold",size=12), axis.text.y = element_text(face="bold", angle=-90,size=10),axis.text.x = element_blank(),axis.title = element_blank()) + guides(color = guide_legend(override.aes = list(size=5)))
# ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/alpha/kraken2_alpha.png")
alpha_diversity_kraken2
```

###Beta Diversity (between samples)
https://grunwaldlab.github.io/analysis_of_microbiome_community_data_in_r/07--diversity_stats.html
- Bray–Curtis: The sum of lesser counts for species present in both communities divided by the sum of all counts in both communities. This can be thought of as a quantitative version of the Sørensen index.
- Weighted Unifrac: The fraction of the phylogenetic tree branch lengths shared by the two communities, weighted by the counts of organisms, so more abundant organisms have a greater influence.
```{r}

# no margin option like alpha diversity, use transpose for samples as colnames
# bray-curtis
beta_dist_kraken2 <-vegdist(t(otu_tab), index="bray") # pairwise comparison between samples

# plot beta diversity using ordination, a way to display high dimensional data
# capture the information in many dimensions by in a smaller number of "artifical" dimensions
mds <- metaMDS(beta_dist_kraken2)
mds_kraken2 <- as.data.frame(mds$points)
mds_kraken2$samples <- rownames(mds_kraken2)
meta_sample$samples <- rownames(meta_sample)
mds_kraken2_meta <- left_join(meta_sample,mds_kraken2, by="samples")

ggplot(mds_kraken2_meta,aes(x=MDS1,y=MDS2,color=Tissue))+
  geom_point(size=5) +
  geom_label_repel(aes(label=samples),
                  point.padding = 0.5)+
  theme_classic()+
  theme(axis.title=element_blank(), axis.text=element_text(size = 10, face = "bold"), legend.text = element_text(size=13), legend.title = element_text(size=14, face="bold"),legend.position = "none",plot.margin = unit(c(0.2,2,0.2,0.2),"cm"))
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/beta/kraken2_beta.png")
mds_kraken2_meta
```

### Differential Abundant with Deseq2
- load DESeq2 package
```{r}
library("DESeq2")

```

- differentially abundant analysis with deseq2
```{r}
desseq_kraken2 <- phyloseq_to_deseq2(cus_kraken2_physeq, ~ Tissue)
desseq_kraken2
kraken_deseq <- DESeq(desseq_kraken2, test="Wald", fitType="parametric")

# pairwise comparison between 2 tissues
Kidney_lung <- results(kraken_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Lung"))
kidney_spleen <- results(kraken_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Spleen"))
lung_spleen <- results(kraken_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Lung","Spleen"))

alpha = 0.01 # significance threhold

# change pairwise comparison results for comparing two different tissues
sigtax_kraken2 <- Kidney_lung[which(Kidney_lung$padj < alpha),] # get tax below the significance thresholf
sigtax_kraken2$Genus <- rownames(sigtax_kraken2) # add a column so genus tax can combine with upper lineages

# add lineages to significant taxa
sigtax_kraken2_lineage <- left_join(as.data.frame(sigtax_kraken2), tax)
dim(sigtax_kraken2_lineage)

```

- Plot significantly abundant taxa from the samples
```{r}
# find the maximum log2fold change from deseq2 output for each cell (tapply)
x<-tapply(sigtax_kraken2_lineage$log2FoldChange,sigtax_kraken2_lineage$Phylum,function(x)max(x))
x<- sort(x,TRUE) # sort these identified phyla in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
sigtax_kraken2_lineage$Phylum <- factor(as.character(sigtax_kraken2_lineage$Phylum),levels = c("Spirochaetes","Proteobacteria","Tenericutes","Firmicutes","Bacteroidetes","Cyanobacteria"))
sig_phylum_color <-
  c("#F8766D","#D89000", "#85AD00" ,"#00A5FF","#DC71FA","#39B600")
kraken2_sig_phylum <- x
# perform the same ordering procedure with genus too
x<-tapply(sigtax_kraken2_lineage$log2FoldChange,sigtax_kraken2_lineage$Genus,function(x)max(x))
x<- sort(x,TRUE) # sort these identified genus in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
sigtax_kraken2_lineage$Genus <- factor(as.character(sigtax_kraken2_lineage$Genus),levels = names(x))
kraken2_sig_genus <- x
ggplot(sigtax_kraken2_lineage, aes(x=Genus, y=log2FoldChange, color= Phylum))+ geom_point(size=5) + 
  theme_classic()+
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5,size = 12, face="bold"),legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold",size=10),axis.title = element_blank())+
  scale_color_manual(values=sig_phylum_color)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/sigtax/kraken2_kidney_lung.png")
```

---

# Clark
## read in all the clark sample's data
```{r}
custom_path="/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/"
species_path <- paste(custom_path, "species_custom/" , sep="")
files <- list.files(species_path)


i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record  %>% select("Lineage","X","X.4","Count") %>% group_by(X.4) %>% mutate(Reads=sum(Count)) %>% select(!c(Count))
  taxa_count <- unique(taxa_count)

  colnames(taxa_count) <- c("Domain","Phylum","Genus",sample_name)
  
  taxa_count %<>% subset(Genus != "Homo" & Genus != "Rattus" & Genus != "")
  rownames(taxa_count) <- NULL # reset row number
 
  if (i == 0){
    genus_full_table <- taxa_count
  } else {
    genus_full_table <- full_join(genus_full_table, taxa_count,by = c("Domain","Phylum" ,"Genus"))
  }
  i = i + 1
}

genus_full_table[is.na(genus_full_table)] <- 0

```

# put clark data into phyloseq object
```{r}
# OTU table
otu_data <- genus_full_table %>% select("Genus","R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S") %>% data.frame
rownames(otu_data)<-genus_full_table$Genus
otu_data <- otu_data[,-1]
OTU <- otu_table(otu_data,taxa_are_rows = T)

# lineage table
tax_data <- genus_full_table %>% 
  select("Domain","Phylum","Genus") %>% data.frame
rownames(tax_data) <- tax_data$Genus
TAX <- tax_table(as.matrix(tax_data))

# add sample's metadata
meta_sample <- read.csv("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/kraken2_sample_meta.csv")
row.names(meta_sample) <- meta_sample$X
meta_sample <- meta_sample[,-1]
sample_meta <- sample_data(meta_sample)
sample_meta

# merge all information together
cus_clark_physeq <- phyloseq(OTU,TAX,sample_meta)
```

### Alpha Diversity (Within Sample diversity)
```{r}

# margin = 2, means samples are the col names
alpha_diversity_clark <- data.frame(Shannon = diversity(otu_data, MARGIN =2, index="shannon"), Simpson = diversity(otu_data, MARGIN =2, index="simpson"), Inverse_Simpson = diversity(otu_data, MARGIN =2, index="invsimpson")) 

# plot Alpha diversity 
plot_richness(cus_clark_physeq, x= "Tissue", color="Subject",measures = c("Shannon", "Simpson", "InvSimpson")) +
  geom_point(size=5) +
   theme_classic()+
  theme(legend.key.size = unit(1.0, "cm"), legend.text = element_text(size=13),legend.title = element_text(size=14,face="bold"), strip.text = element_text(face="bold",size=12), axis.text.y = element_text(face="bold", angle=-90,size=10),axis.text.x=element_blank(),axis.title = element_blank()) + guides(color = guide_legend(override.aes = list(size=5)))
# ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/alpha/clark_alpha.png")
alpha_diversity_clark
```

###Beta Diversity (between samples)
```{r}

# no margin option like alpha diversity, use transpose for samples as colnames
# bray-curtis
beta_dist_clark <-vegdist(t(otu_data), index="bray") # pairwise comparison between samples

# plot beta diversity using ordination, a way to display high dimensional data
# capture the information in many dimensions by in a smaller number of "artifical" dimensions
mds <- metaMDS(beta_dist_clark)
mds_clark <- as.data.frame(mds$points)
mds_clark$samples <- rownames(mds_clark)
meta_sample$samples <- rownames(meta_sample)
mds_clark_meta <- left_join(meta_sample,mds_clark, by="samples")

ggplot(mds_clark_meta,aes(x=MDS1,y=MDS2,color=Tissue))+
  geom_point(size=5) +
  geom_label_repel(aes(label=samples),
                  point.padding = 0.5)+
  theme_classic()+
  theme(axis.title=element_blank(), axis.text=element_text(size = 10, face = "bold"), legend.text = element_text(size=13), legend.title = element_text(size=14, face="bold"),legend.position = "none",plot.margin = unit(c(0.2,2,0.2,0.2),"cm"))
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/beta/clark_beta.png")
```

### Differential Abundant with Deseq2

- differentially abundant analysis with deseq2
```{r}
desseq_clark <- phyloseq_to_deseq2(cus_clark_physeq, ~ Tissue)
desseq_clark
clark_deseq <- DESeq(desseq_clark, test="Wald", fitType="parametric")

# pairwise comparison between 2 tissues
Kidney_lung <- results(clark_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Lung"))
kidney_spleen <- results(clark_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Spleen"))
lung_spleen <- results(clark_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Lung","Spleen"))

alpha = 0.01 # significance threhold

# change pairwise comparison results for comparing two different tissues
sigtax_clark <- Kidney_lung[which(Kidney_lung$padj < alpha),] # get tax below the significance thresholf
sigtax_clark$Genus <- rownames(sigtax_clark) # add a column so genus tax can combine with upper lineages

# add lineages to significant taxa
clark_deseq_lineage <- left_join(as.data.frame(sigtax_clark), tax_data)
dim(clark_deseq_lineage)

```

- Plot significantly abundant taxa from the samples
```{r}
# find the maximum log2fold change from deseq2 output for each cell (tapply)
x<-tapply(clark_deseq_lineage$log2FoldChange,clark_deseq_lineage$Phylum,function(x)max(x))
x<- sort(x,TRUE) # sort these identified phyla in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
clark_deseq_lineage$Phylum <- factor(as.character(clark_deseq_lineage$Phylum),levels = c("Spirochaetes","Proteobacteria","Tenericutes","Firmicutes","Bacteroidetes","Peploviricota","Cyanobacteria"))
clark_sig_phylum <- x
# perform the same ordering procedure with genus too
x<-tapply(clark_deseq_lineage$log2FoldChange,clark_deseq_lineage$Genus,function(x)max(x))
x<- sort(x,TRUE) # sort these identified genus in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
clark_deseq_lineage$Genus <- factor(as.character(clark_deseq_lineage$Genus),levels = names(x))
clark_sig_genus <- x
sig_phylum_color <-
  c("#F8766D","#D89000", "#85AD00" ,"#00A5FF","#DC71FA","#00C0B8","#39B600")
ggplot(clark_deseq_lineage, aes(x=Genus, y=log2FoldChange, color= Phylum))+ geom_point(size=5) + 
  theme_classic()+
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5,size=12,face="bold"),legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold",size=10),axis.title = element_blank()) +
  scale_color_manual(values=sig_phylum_color)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/sigtax/clark_kidney_lung.png")
```
 
---

# Clark-s 
## read in all the clarks sample's data
```{r}
custom_path="/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/clark/"
species_path <- paste(custom_path, "species_rat_spaced/" , sep="")
files <- list.files(species_path)


i=0
for (file in files){
  sample_name <- unlist(strsplit(file, ".",fixed=T))[1]
  sample_record <- read.csv(paste(species_path, file,sep=""), header = TRUE)
  # get the domain column
  taxa_count <- sample_record  %>% select("Lineage","X","X.4","Count") %>% group_by(X.4) %>% mutate(Reads=sum(Count)) %>% select(!c(Count))
  taxa_count <- unique(taxa_count)

  colnames(taxa_count) <- c("Domain","Phylum","Genus",sample_name)
  
  taxa_count %<>% subset(Genus != "Homo" & Genus != "Rattus" & Genus != "")
  rownames(taxa_count) <- NULL # reset row number
 
  if (i == 0){
    genus_full_table <- taxa_count
  } else {
    genus_full_table <- full_join(genus_full_table, taxa_count,by = c("Domain","Phylum" ,"Genus"))
  }
  i = i + 1
}

genus_full_table[is.na(genus_full_table)] <- 0
genus_full_table
```

# put clarks data into phyloseq object
```{r}
# OTU table
otu_data <- genus_full_table %>% select("Genus","R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S") %>% data.frame
rownames(otu_data)<-genus_full_table$Genus
otu_data <- otu_data[,-1]
OTU <- otu_table(otu_data,taxa_are_rows = T)

# lineage table
tax_data <- genus_full_table %>% 
  select("Domain","Phylum","Genus") %>% data.frame
rownames(tax_data) <- tax_data$Genus
TAX <- tax_table(as.matrix(tax_data))

# add sample's metadata
meta_sample <- read.csv("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/kraken2_sample_meta.csv")
row.names(meta_sample) <- meta_sample$X
meta_sample <- meta_sample[,-1]
sample_meta <- sample_data(meta_sample)
sample_meta

# merge all information together
cus_clarks_physeq <- phyloseq(OTU,TAX,sample_meta)
```

### Alpha Diversity (Within Sample diversity)
```{r}
# margin = 2, means samples are the col names
alpha_diversity_clarks <- data.frame(Shannon = diversity(otu_data, MARGIN =2, index="shannon"), Simpson = diversity(otu_data, MARGIN =2, index="simpson"), Inverse_Simpson = diversity(otu_data, MARGIN =2, index="invsimpson")) 
alpha_diversity_clarks
# plot Alpha diversity 
plot_richness(cus_clarks_physeq, x= "Tissue", color="Subject",measures = c("Shannon", "Simpson", "InvSimpson")) +
  geom_point(size=5) +
   theme_classic()+
  theme(legend.key.size = unit(1.0, "cm"), legend.text = element_text(size=13),legend.title = element_text(size=14,face="bold"), strip.text = element_text(face="bold",size=12), axis.text.y  = element_text(face="bold", angle=-90,size=10),axis.text.x=element_blank(),axis.title = element_blank()) + guides(color = guide_legend(override.aes = list(size=5)))
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/alpha/clarks_alpha.png")

```

###Beta Diversity (between samples)
```{r}

# no margin option like alpha diversity, use transpose for samples as colnames
# bray-curtis
beta_dist_clarks <-vegdist(t(otu_data), index="bray") # pairwise comparison between samples

# plot beta diversity using ordination, a way to display high dimensional data
# capture the information in many dimensions by in a smaller number of "artifical" dimensions
mds <- metaMDS(beta_dist_clarks)
mds_clarks <- as.data.frame(mds$points)
mds_clarks$samples <- rownames(mds_clarks)
meta_sample$samples <- rownames(meta_sample)
mds_clarks_meta <- left_join(meta_sample,mds_clarks, by="samples")

ggplot(mds_clarks_meta,aes(x=MDS1,y=MDS2,color=Tissue))+
 geom_point(size=5) +
  geom_label_repel(aes(label=samples),
                  point.padding = 0.5)+
  theme_classic()+
  theme(axis.title=element_blank(), axis.text=element_text(size = 10, face = "bold"), legend.text = element_text(size=13), legend.title = element_text(size=14, face="bold"),legend.position = "none",plot.margin = unit(c(0.2,2,0.2,0.2),"cm"))
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/beta/clarks_beta.png")
```

### Differential Abundant with Deseq2

- differentially abundant analysis with deseq2
```{r}
desseq_clarks <- phyloseq_to_deseq2(cus_clarks_physeq, ~ Tissue)
desseq_clarks
clarks_deseq <- DESeq(desseq_clarks, test="Wald", fitType="parametric")

# pairwise comparison between 2 tissues
Kidney_lung <- results(clarks_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Lung"))
kidney_spleen <- results(clarks_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Kidney","Spleen"))
lung_spleen <- results(clarks_deseq, cooksCutoff = FALSE,contrast = c("Tissue","Lung","Spleen"))

alpha = 0.01 # significance threhold

# change pairwise comparison results for comparing two different tissues
sigtax_clarks <- Kidney_lung[which(Kidney_lung$padj < alpha),] # get tax below the significance thresholf
sigtax_clarks$Genus <- rownames(sigtax_clarks) # add a column so genus tax can combine with upper lineages

# add lineages to significant taxa
clarks_deseq_lineage <- left_join(as.data.frame(sigtax_clarks), tax_data)
dim(clarks_deseq_lineage)


```

- Plot significantly abundant taxa from the samples
```{r}
# find the maximum log2fold change from deseq2 output for each cell (tapply)
x<-tapply(clarks_deseq_lineage$log2FoldChange,clarks_deseq_lineage$Phylum,function(x)max(x))
x<- sort(x,TRUE) # sort these identified phyla in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
clarks_deseq_lineage$Phylum <- factor(as.character(clarks_deseq_lineage$Phylum),levels = c("Spirochaetes","Proteobacteria","Tenericutes","Firmicutes","Bacteroidetes","Peploviricota","Cyanobacteria"))
clarks_sig_phylum <- x
# perform the same ordering procedure with genus too
x<-tapply(clarks_deseq_lineage$log2FoldChange,clarks_deseq_lineage$Genus,function(x)max(x))
x<- sort(x,TRUE) # sort these identified genus in descending order
# order the sig taxa base on the decreasing order for the fold of changes between tissues
clarks_deseq_lineage$Genus <- factor(as.character(clarks_deseq_lineage$Genus),levels = names(x))
clarks_sig_genus <- x
ggplot(clarks_deseq_lineage, aes(x=Genus, y=log2FoldChange, color= Phylum))+ geom_point(size=5) + 
  theme_classic()+
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5,size=12,face="bold"),legend.key.size = unit(0.6, "cm"), legend.text = element_text(size=12),legend.title = element_text(size=12,face="bold"), axis.text.y = element_text(face="bold", size=10),axis.title = element_blank()) +
  scale_color_manual(values=sig_phylum_color)
ggsave("/Users/rachel/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/plots/sigtax/clarks_kidney_lung.png")


```

# OVERVIEW: Comparison Between Three Softwares

### sig different genus kidney vs lung
```{r}
# kraken2
cat("Kraken2:\n")
kraken2_sig_genus
sig_genus_kraken<- rownames(as.data.frame(kraken2_sig_genus))

# clark
cat("Clark:\n")
clark_sig_genus
sig_genus_clark<- rownames(as.data.frame(clark_sig_genus))



#clark-s
cat("Clark-s:\n")
clarks_sig_genus
sig_genus_clarks<- rownames(as.data.frame(clarks_sig_genus))


intersect(intersect(sig_genus_kraken, sig_genus_clark),sig_genus_clarks)

```
### sig different phylum kidney vs lung
```{r}
# kraken2
cat("Kraken2:\n")
kraken2_sig_phylum

# clark
cat("Clark:\n")
clarks_sig_phylum


#clark-s
cat("Clark-s:\n")
clark_sig_phylum

```
### Alpha Diversity
```{r}
# kraken2
cat("Kraken2:\n")
alpha_diversity_kraken2

# clark
cat("Clark:\n")
alpha_diversity_clark


#clark-s
cat("Clark-s:\n")
clark_sig_phylum
alpha_diversity_clarks

```

## if the alpha diversity provided for each sample with different software has a significant difference (Friedman Test)

### Shannon
```{r}
# friedman test
# paired test for more than two paired samples that is not normally distributed
samples <- rep(rownames(alpha_diversity_kraken2),3)

three_alpha <- data.frame(sample_id = samples, software=c(rep("kraken",12),rep("clark",12), rep("clark-s",12)), alpha_index=c(alpha_diversity_kraken2$Shannon,alpha_diversity_clark$Shannon,alpha_diversity_clarks$Shannon))
three_alpha$software <- factor(three_alpha$software, levels=c("kraken","clark","clark-s"))
three_alpha$sample_id <- factor(three_alpha$sample_id, levels = rownames(alpha_diversity_kraken2))
# The null hypothesis is that apart from an effect of blocks, the location parameter of y is the same in each of the groups
friedman.test(alpha_index ~ sample_id | software, data = three_alpha) # (values ~ group | block)

################# wilcoxon test for pair-wise significance test #######################

# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("kraken2",12),rep("clark",12)),alpha_index=c(alpha_diversity_kraken2$Shannon,alpha_diversity_clark$Shannon))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("kraken2","clark"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected

# wilcoxon test for pair-wise significance test
# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("clark",12),rep("clarks",12)),alpha_index=c(alpha_diversity_clark$Shannon,alpha_diversity_clarks$Shannon))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("clark","clarks"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected

# wilcoxon test for pair-wise significance test
# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("kraken2",12),rep("clarks",12)),alpha_index=c(alpha_diversity_kraken2$Shannon,alpha_diversity_clarks$Shannon))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("kraken2","clarks"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected
```

### Simpson
```{r}
# friedman test
# paired test for more than two paired samples that is not normally distributed
samples <- rep(rownames(alpha_diversity_kraken2),3)

three_alpha <- data.frame(sample_id = samples, software=c(rep("kraken",12),rep("clark",12), rep("clark-s",12)), alpha_index=c(alpha_diversity_kraken2$Simpson,alpha_diversity_clark$Simpson,alpha_diversity_clarks$Simpson))
three_alpha$software <- factor(three_alpha$software, levels=c("kraken","clark","clark-s"))
three_alpha$sample_id <- factor(three_alpha$sample_id, levels = rownames(alpha_diversity_kraken2))
# The null hypothesis is that apart from an effect of blocks, the location parameter of y is the same in each of the groups
friedman.test(alpha_index ~ sample_id | software, data = three_alpha) # (values ~ group | block)

################# wilcoxon test for pair-wise significance test #######################

# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("kraken2",12),rep("clark",12)),alpha_index=c(alpha_diversity_kraken2$Simpson,alpha_diversity_clark$Simpson))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("kraken2","clark"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected

# wilcoxon test for pair-wise significance test
# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("clark",12),rep("clarks",12)),alpha_index=c(alpha_diversity_clark$Simpson,alpha_diversity_clarks$Simpson))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("clark","clarks"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected

# wilcoxon test for pair-wise significance test
# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("kraken2",12),rep("clarks",12)),alpha_index=c(alpha_diversity_kraken2$Simpson,alpha_diversity_clarks$Simpson))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("kraken2","clarks"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected
```

### Inverse Simpson
```{r}
# friedman test
# paired test for more than two paired samples that is not normally distributed
samples <- rep(rownames(alpha_diversity_kraken2),3)

three_alpha <- data.frame(sample_id = samples, software=c(rep("kraken",12),rep("clark",12), rep("clark-s",12)), alpha_index=c(alpha_diversity_kraken2$Inverse_Simpson,alpha_diversity_clark$Inverse_Simpson,alpha_diversity_clarks$Inverse_Simpson))
three_alpha$software <- factor(three_alpha$software, levels=c("kraken","clark","clark-s"))
three_alpha$sample_id <- factor(three_alpha$sample_id, levels = rownames(alpha_diversity_kraken2))
# The null hypothesis is that apart from an effect of blocks, the location parameter of y is the same in each of the groups
friedman.test(alpha_index ~ sample_id | software, data = three_alpha) # (values ~ group | block)

################# wilcoxon test for pair-wise significance test #######################

# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("kraken2",12),rep("clark",12)),alpha_index=c(alpha_diversity_kraken2$Inverse_Simpson,alpha_diversity_clark$Inverse_Simpson))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("kraken2","clark"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected

# wilcoxon test for pair-wise significance test
# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("clark",12),rep("clarks",12)),alpha_index=c(alpha_diversity_clark$Inverse_Simpson,alpha_diversity_clarks$Inverse_Simpson))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("clark","clarks"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected

# wilcoxon test for pair-wise significance test
# wilcoxon, kraken2 vs clark
samples <- rep(rownames(alpha_diversity_kraken2),2)
two_alpha <- data.frame(sample_id= samples, software=c(rep("kraken2",12),rep("clarks",12)),alpha_index=c(alpha_diversity_kraken2$Inverse_Simpson,alpha_diversity_clarks$Inverse_Simpson))
two_alpha$sample_id <- factor(two_alpha$sample_id, levels =unique(samples))
two_alpha$software <- factor(two_alpha$software, levels = c("kraken2","clarks"))
wilcox.test(two_alpha$alpha_index ~ two_alpha$software, paired = TRUE) # no rejected
```

### Beta_Diversity

```{r}
# kraken2
cat("Kraken2:\n")
beta_dist_kraken2

# clark
cat("Clark:\n")
beta_dist_clark


#clark-s
cat("Clark-s:\n")
clark_sig_phylum
beta_dist_clarks

```

### Friedman test for beta diversities (bray indices) from Kraken2, Clark and Clark-S
```{r}
# friedman test
# paired test for more than two paired samples that is not normally distributed
between_samples <- as.character(rep(seq(1:66),3)) # use number as index to represent pairwise relationships

# convert distance matrix to vector from top to bottom row by row
beta_dist_kraken2_matrix <- as.matrix(beta_dist_kraken2)
kraken2_beta_vector <- beta_dist_kraken2_matrix[lower.tri(beta_dist_kraken2_matrix)][order(row(beta_dist_kraken2_matrix)[lower.tri(row(beta_dist_kraken2_matrix))])]
# convert distance matrix to vector from top to bottom row by row
beta_dist_clark_matrix <- as.matrix(beta_dist_clark)
clark_beta_vector <- beta_dist_clark_matrix[lower.tri(beta_dist_clark_matrix)][order(row(beta_dist_clark_matrix)[lower.tri(row(beta_dist_clark_matrix))])]
# convert distance matrix to vector from top to bottom row by row
beta_dist_clarks_matrix <- as.matrix(beta_dist_clarks)
clarks_beta_vector <- beta_dist_clarks_matrix[lower.tri(beta_dist_clarks_matrix)][order(row(beta_dist_clarks_matrix)[lower.tri(row(beta_dist_clarks_matrix))])]

three_beta <- data.frame(compare_id = between_samples, software=c(rep("kraken",66),rep("clark",66), rep("clark-s",66)), beta_index=c(kraken2_beta_vector,clark_beta_vector,clarks_beta_vector))
three_beta$software <- factor(three_beta$software, levels=c("kraken","clark","clark-s"))
three_beta$compare_id <- factor(three_beta$compare_id, levels = as.character(seq(1:66)))
# The null hypothesis is that apart from an effect of blocks, the location parameter of y is the same in each of the groups
friedman.test(beta_index ~ compare_id | software, data = three_beta) # (values ~ group | block)

################# wilcoxon test for pair-wise significance test #######################

between_samples <- as.character(rep(seq(1:66),2)) # use number as index to represent pairwise relationships
# wilcoxon, kraken2 vs clark

two_beta <- data.frame(sample_id= between_samples, software=c(rep("kraken2",66),rep("clark",66)),beta_index=c(kraken2_beta_vector,clark_beta_vector))
two_beta$sample_id <- factor(two_beta$sample_id, levels =as.character(seq(1:66)))
two_beta$software <- factor(two_beta$software, levels = c("kraken2","clark"))
wilcox.test(two_beta$beta_index ~ two_beta$software, paired = TRUE) # no rejected

# wilcoxon test for pair-wise significance test
# wilcoxon, kraken2 vs clark

two_beta <- data.frame(sample_id= between_samples, software=c(rep("clark",66),rep("clarks",66)),beta_index=c(clark_beta_vector,clarks_beta_vector))
two_beta$sample_id <- factor(two_beta$sample_id, levels =as.character(seq(1:66)))
two_beta$software <- factor(two_beta$software, levels = c("clark","clarks"))
wilcox.test(two_beta$beta_index ~ two_beta$software, paired = TRUE) # no rejected

# wilcoxon test for pair-wise significance test
# wilcoxon, kraken2 vs clark

two_beta <- data.frame(sample_id= between_samples, software=c(rep("kraken2",66),rep("clarks",66)),beta_index=c(kraken2_beta_vector,clarks_beta_vector))
two_beta$sample_id <- factor(two_beta$sample_id, levels =as.character(seq(1:66)))
two_beta$software <- factor(two_beta$software, levels = c("kraken2","clarks"))
wilcox.test(two_beta$beta_index ~ two_beta$software, paired = TRUE) # no rejected


```
