---
title: "kraken2_db_comparison"
output: html_document
---

```{r}
path <- "/Users/rx32940/Dropbox/5.Rachel-projects/Metagenomic_Analysis/final_analysis/kraken2/"
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyverse)
library(plotly)
library(coin)
library(rstatix)
```

# MiniKraken domain 
```{r}

mini_path <- paste(path,"minikraken/absolute/",sep="")
files <- list.files(mini_path)

mini_path
i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(mini_path,file,sep=""),sep="\t",header=FALSE)
  # add "| trimws(V6) == "other sequences"" to subset if counting total number of reads
  domain_absolute <- current_sample %>% subset(V4 == "D" | V4 == "U" | trimws(V6) == "other sequences") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  
  colnames(domain_absolute) <- c("Domain", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    domain_all_samples <-  domain_absolute
    
  } else {
    domain_all_samples <- full_join(domain_all_samples, domain_absolute,by="Domain")
    
  }
  i = i + 1
}

# save(domain_all_samples, file ="minikraken_absolute_domain.Rdata")

domain_all_samples[is.na(domain_all_samples)] <- 0
keys_to_gather <- colnames(domain_all_samples)[-1] # gather samples, exclude "Domain" from colnames

# prepare dataframe for ggplot
domain_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

order <- unique(domain_all_samples$samples)
domain_all_samples$samples <- factor(domain_all_samples$samples, levels = order)
domain_all_samples$Domain <- factor(domain_all_samples$Domain, level=unique(domain_all_samples$Domain))


# Total number of reads for each sample (unclassified + three domains + other sequences)
domain_all_samples %>% group_by(samples) %>% summarise(total_reads=sum(Reads)) -> mini_total_reads
as.data.frame(list(mean_reads= mean(mini_total_reads$total_reads),std=sd(mini_total_reads$total_reads))) -> mini_total_reads_summary# avg number of reads across sample

# other sequences summary
domain_all_samples %>% subset(Domain == "other sequences") -> mini_other_sequences
mini_other_sequences %>%summarise(mean_other=mean(Reads),std=sd(Reads)) -> mini_other_sequences_summary

# unclassified reads summary
domain_all_samples %>% subset(Domain == "unclassified") -> mini_unc
mini_unc %>%summarise(mean_other=mean(Reads),std=sd(Reads)) -> mini_unc_summary

# microbial reads summary
domain_all_samples %>% subset(Domain == "Eukaryota" | Domain == "Bacteria" | Domain == "Viruses" | Domain == "Archaea") -> microbial_mini

unclass_mini <- domain_all_samples %>% subset(Domain == "unclassified")
other_mini <- domain_all_samples %>% subset(Domain == "other sequences")
bacteria_mini <- domain_all_samples %>% subset(Domain == "Bacteria" )
eukaryota_mini <- domain_all_samples %>% subset(Domain == "Eukaryota")
viruses_mini <- domain_all_samples %>% subset(Domain == "Viruses")
archaea_mini <- domain_all_samples %>% subset(Domain == "Archaea")

microbial_mini %>% group_by(samples) %>% summarise(total_microbial=sum(Reads)) -> microbial_mini_summary


mini_total_reads_summary# avg number of reads across sample
mini_other_sequences_summary # avg number of reads classified into "other sequences"
mini_unc_summary
microbial_mini_summary %>% summarise(mean= mean(total_microbial), std= sd(total_microbial))
microbial_mini_summary
```

# standard Kraken2 domain
```{r}

std_path <- paste(path,"standard/absolute/",sep="")
files <- list.files(std_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(std_path,file,sep=""),sep="\t",header=FALSE)
  # add "| trimws(V6) == "other sequences"" to subset if counting total number of reads
  domain_absolute <- current_sample %>% subset(V4 == "D" | V4 == "U" | trimws(V6) == "other sequences") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))

  colnames(domain_absolute) <- c("Domain", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    domain_all_samples <-  domain_absolute

  } else {
    domain_all_samples <- full_join(domain_all_samples, domain_absolute,by="Domain")

  }
  i = i + 1
}

# save(domain_all_samples, file ="standard_absolute_domain.Rdata")

domain_all_samples[is.na(domain_all_samples)] <- 0
keys_to_gather <- colnames(domain_all_samples)[-1] # gather samples, exclude "Domain" from colnames

# prepare dataframe for ggplot
domain_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

order <- unique(domain_all_samples$samples)
domain_all_samples$samples <- factor(domain_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
domain_all_samples$Domain <- factor(domain_all_samples$Domain, level=unique(domain_all_samples$Domain))


# test_sample <- read.csv(paste(absolute_path,"R22_K.kreport",sep=""),sep="\t",header=FALSE)
# test_absolute <- test_sample %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V4,V2,V1)) %>% subset(V4 == "D"| V4 == "U" | V4 == "R" | V4 == "R1"|V4=="G")  %>% group_by(V4) %>% summarise(sum_read= sum(V2), sum_percent=sum(V1))

# Total number of reads for each sample (unclassified + three domains + other sequences)
domain_all_samples %>% group_by(samples) %>% summarise(total_reads=sum(Reads)) -> std_total_reads
as.data.frame(list(mean_reads=mean(std_total_reads$total_reads),std=sd(std_total_reads$total_reads))) -> std_total_reads_summary# avg number of reads across sample

# other sequences summary
domain_all_samples %>% subset(Domain == "other sequences") -> other_sequences_std
other_sequences_std %>% summarise(mean_other=mean(Reads),std=sd(Reads)) -> other_seq_std_summary

# unclassified summary
domain_all_samples %>% subset(Domain == "unclassified") -> std_unc
std_unc %>%summarise(mean_other=mean(Reads),std=sd(Reads)) -> std_unc_summary

# microbial reads summary
domain_all_samples %>% subset(Domain == "Eukaryota" | Domain == "Bacteria" | Domain == "Viruses" | Domain == "Archaea") -> microbial_std
unclass_std <- domain_all_samples %>% subset(Domain == "unclassified")
other_std <- domain_all_samples %>% subset(Domain == "other sequences")
bacteria_std <- domain_all_samples %>% subset(Domain == "Bacteria" )
eukaryota_std <- domain_all_samples %>% subset(Domain == "Eukaryota")
viruses_std <- domain_all_samples %>% subset(Domain == "Viruses")
archaea_std <- domain_all_samples %>% subset(Domain == "Archaea")

microbial_std %>% group_by(samples) %>% summarise(total_microbial=sum(Reads)) -> microbial_std_summary

std_total_reads_summary
other_seq_std_summary
std_unc_summary
microbial_std_summary %>% summarise(mean= mean(total_microbial), std= sd(total_microbial))
```

# Custom db domain 
```{r}

cus_path = paste(path,"custom/absolute/",sep="")
files <- list.files(cus_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(cus_path,file,sep=""),sep="\t",header=FALSE)
  # add "| trimws(V6) == "other sequences"" to subset if counting total number of reads
  domain_absolute <- current_sample %>% subset(V4 == "D" | V4 == "U" | trimws(V6) == "other sequences") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))

  colnames(domain_absolute) <- c("Domain", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    domain_all_samples <-  domain_absolute

  } else {
    domain_all_samples <- full_join(domain_all_samples, domain_absolute,by="Domain")

  }
  i = i + 1
}

# save(domain_all_samples, file ="custom_absolute_domain.Rdata")

domain_all_samples[is.na(domain_all_samples)] <- 0
keys_to_gather <- colnames(domain_all_samples)[-1] # gather samples, exclude "Domain" from colnames

# prepare dataframe for ggplot
domain_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")


domain_all_samples$samples <- factor(domain_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
domain_all_samples$Domain <- factor(domain_all_samples$Domain, level=unique(domain_all_samples$Domain))

# Total number of reads for each sample (unclassified + three domains + other sequences)
domain_all_samples %>% group_by(samples) %>% summarise(total_reads=sum(Reads)) -> cus_total_reads
as.data.frame(list(mean_reads=mean(cus_total_reads$total_reads),std=sd(cus_total_reads$total_reads))) -> cus_total_reads_summary # avg number of reads across sample

# other sequences summary
domain_all_samples %>% subset(Domain == "other sequences") -> cus_other_seq
cus_other_seq %>% summarise(mean_other=mean(Reads),std=sd(Reads)) -> other_seq_cus_summary

# unclassified reads summary
domain_all_samples %>% subset(Domain == "unclassified") -> cus_unc
cus_unc %>%summarise(mean_other=mean(Reads),std=sd(Reads)) -> cus_unc_summary

# microbial reads summary
domain_all_samples %>% subset(Domain == "Eukaryota" | Domain == "Bacteria" | Domain == "Viruses" | Domain == "Archaea") -> microbial_cus

unclass_cus <- domain_all_samples %>% subset(Domain == "unclassified")
other_cus <- domain_all_samples %>% subset(Domain == "other sequences")
bacteria_cus <- domain_all_samples %>% subset(Domain == "Bacteria" )
eukaryota_cus <- domain_all_samples %>% subset(Domain == "Eukaryota")
viruses_cus <- domain_all_samples %>% subset(Domain == "Viruses")
archaea_cus <- domain_all_samples %>% subset(Domain == "Archaea")

microbial_cus %>% group_by(samples) %>% summarise(total_microbial=sum(Reads)) -> microbial_cus_summary

cus_total_reads_summary # avg number of reads across sample
other_seq_cus_summary
cus_unc_summary
microbial_cus_summary %>% summarise(mean= mean(total_microbial), std= sd(total_microbial))

```
# Combine three databases

```{r}

kraken2_domain_list <- c("unclass", "other", "eukaryota", "bacteria", "viruses", "archaea")

combined_kraken2_db <- df <- data.frame(Date=as.Date(character()),
                 File=character(), 
                 User=character(), 
                 stringsAsFactors=FALSE) 
for (d in kraken2_domain_list){
  join1 <- left_join(get(paste0(d,"_mini")), get(paste0(d,"_std")), by= c("Domain", "samples"))
  combined_temp <- left_join(join1, get(paste0(d,"_cus")), by= c("Domain", "samples"))
  combined_temp %<>% dplyr::rename("MiniKraken2" = Reads.x, "Standard" = Reads.y, "Customized" = Reads)
  combined_kraken2_db <- rbind(combined_kraken2_db, combined_temp)
}

combined_kraken2_db_longer <- combined_kraken2_db %>% pivot_longer(c("MiniKraken2","Standard", "Customized"),names_to = "Kraken2 Database Comparison", values_to = "Reads")
#combined_kraken2_db_longer %>% 

#reorder database
combined_kraken2_db_longer$`Kraken2 Database Comparison` <- factor(combined_kraken2_db_longer$`Kraken2 Database Comparison`, levels = c("MiniKraken2","Standard", "Customized"))
#capitalize
combined_kraken2_db_longer$Domain <- str_to_title(combined_kraken2_db_longer$Domain)
combined_kraken2_db_longer$Domain <- factor(combined_kraken2_db_longer$Domain, levels = c("Unclassified", "Other Sequences", "Eukaryota","Bacteria", "Viruses","Archaea"))

library(ggsignif)
library(scales)

combined_kraken2_db_longer
### plot 
# plot_kraken2_p <- ggplot(combined_kraken2_db_longer,aes(x= `Kraken2 Database Comparison`, y= `Reads`))+
#       geom_boxplot(color="grey",alpha=0.1)+
#   scale_shape_manual(values = c(19,18,17,15,seq(1,12,1)))+
#   geom_point(aes(shape=samples),size=2) +
# 
# stat_signif(comparisons = combn(levels(combined_kraken2_db_longer$`Kraken2 Database Comparison`),2,simplify = F),
#           map_signif_level =c("***"=0.001,"**"=0.01,"*"=0.05, " "=2), test ="wilcox.test", position = "identity",
#           test.args = list(n=18,paired=T, exact=TRUE), vjust=0.6,
#                         textsize=6,
#                         size=0.5,
#                         step_increase = 0.05, na.rm = TRUE, show.legend = TRUE, fontface = "bold")+
#   facet_wrap(vars(Domain), ncol = 3, scales = "free_y")+
#   theme_bw()+
# 
#   theme(axis.text.x = element_text(angle=45, vjust=0.8,hjust = 0.8, size =12),axis.text.y = element_text(size=12), strip.text = element_text(size=12),axis.title = element_text(size=12), legend.text = element_text(size=12),legend.title = element_text(size=12))+
#     scale_y_continuous(labels = comma)+
#   labs(y="Number of Reads")

manual_color <- c("#BFC9CA", "#AED6F1", "#0E6251","#0E6251","#6E2C00","#EDBB99")

manual_color <-c("#D6EAF8","#27AE60","#FAD7A0","#4361EE","#FDFFB6","#CAFFBF","#9BF6FF","#BDB2FF","#FFC6FF","#233D4D","#FE7F2D","#FCCA46","#A1C181",
                 "#619B8A","#EE6352","#59CD90","#3FA7D6","#FAC05E","#F79D84","#50514F","#F25F5C","#FFE066","#247BA0","#70C1B3",
                 "#FFD1E8","#C8A439","#465D6F","#A94303","#708841","#540D6E","#EE4266","#FFD23F","#3BCEAC","#0EAD69","#E3AB00ff","#D1DED3ff",
                 "#AD8EB0ff","#58BADCff","#A71D2Eff","#86D8BBff","#0050ADff","#97CBF0ff","#67A280ff","#F20089","#41EAD4","#FBFF12","#54478C")

combined_kraken2_db_longer_1 <- combined_kraken2_db_longer %>% group_by(samples, `Kraken2 Database Comparison`) %>% mutate(`Percentage of Reads` = Reads/sum(Reads) ) %>% pivot_longer(c("Reads",`Percentage of Reads`), names_to = "Reads.Type", values_to = "values")

combined_kraken2_db_longer_1$samples <- factor(combined_kraken2_db_longer_1$samples, levels=c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
combined_db_p <- ggplot(combined_kraken2_db_longer_1 %>% subset(Reads.Type == "Reads"), aes(x= samples, y=values), ncol=2)+
  geom_bar(stat="identity", aes(fill=Domain))+
  facet_grid(rows = vars(`Kraken2 Database Comparison`), cols = vars(Reads.Type), switch = "y")+
  scale_y_continuous(labels = comma)+
  theme_classic()+
  scale_fill_manual(values = manual_color)+
  theme(axis.text.x = element_text(angle=90,size=12), axis.text = element_text(size=12), axis.title  = element_text(size=12),strip.placement = "outer", axis.title.y = element_blank(), strip.text = element_text(size=12), axis.title.x = element_blank(), panel.spacing = unit(1,"lines"),legend.position="none")#+
  scale_fill_grey(start = 0, end=0.9)
combined_db_p
combined_db_c <- ggplot(combined_kraken2_db_longer_1%>% subset(Reads.Type == "Percentage of Reads") , aes(x= samples, y=values), ncol=2)+
  geom_bar(stat="identity", aes(fill=Domain))+
  facet_grid(rows = vars(`Kraken2 Database Comparison`), cols = vars(Reads.Type))+
  scale_y_continuous(labels = comma)+
  theme_classic()+
  theme(axis.text.x = element_text(angle=90,size=12), axis.text = element_text(size=12) ,axis.title  = element_text(size=12),strip.placement = "outer", axis.title.y = element_blank(), strip.text = element_text(size=12), axis.title.x = element_blank(), panel.spacing = unit(1,"lines"), legend.text = element_text(size=12), legend.title = element_text(size=12))+
  scale_fill_manual(values = manual_color)+
  scale_y_continuous(labels= scales::percent,position="right")#+
  scale_fill_grey(start = 0, end=0.9)
combined_db_c
combined_phylum_software <- grid.arrange(combined_db_p, combined_db_c, ncol = 2)

combined_phylum_software 
ggsave(plot = combined_phylum_software,  dpi = 600, ,width=8,filename = "/Users/rx32940/Desktop/kraken2_db_barplot.pdf")
```

```{r}
# combined_kraken2_db_longer %>% group_by(Domain) %>% summarise(w = wilcox.test(Reads~`Kraken2 Database Comparison`, data=., paired=TRUE))


```


# "other sequences" reads with custom database is significantly different than the mini and std databases reads
- why use **paired Wilcoxon signed rank test**: https://www.nki.nl/media/837516/m343.pdf
- Tutorial for Wilcoxon signed rank test:https://towardsdatascience.com/wilcoxon-test-in-r-how-to-compare-2-groups-under-the-non-normality-assumption-6fb7f9e92400

```{r}

# all three distribution for reads classified into the "other sequences" did not follow a normal distribution with the reject the null hypothesis at 0.05
shapiro.test(mini_other_sequences$Reads)
shapiro.test(cus_other_seq$Reads)
shapiro.test(other_sequences_std$Reads)

# use Wilcoxon test for difference with non-normal distributed groups

# mini vs std for other sequences reads
mini_std_other <- data.frame(db=c(rep("mini",12),rep("std",12)), other_seq_reads=c(mini_other_sequences$Reads,other_sequences_std$Reads))
# wilcoxon test requires exactly 2 levels of groups factors
mini_std_other$db <- factor(mini_std_other$db, levels=c("mini","std"))

# wilcoxon test 
# alternative hypothesis, cus db with significantly lesser reads
# since tested on the same data set with different methods, paired in true
test <- wilcox.test(mini_std_other$other_seq_reads ~ mini_std_other$db, paired =TRUE, exact = TRUE, correction = T)
qnorm(test$p.value/2)
coin::wilcoxsign_test(mini_std_other$other_seq_reads ~ mini_std_other$db, paired=TRUE, distribution = "exact" ,zero.method="Wilcoxon" ) # not rejected
mini_std_other %>% wilcox_effsize(other_seq_reads ~ db, paired = TRUE)


# mini vs custom for other sequences reads
mini_cus_other <- data.frame(db=c(rep("mini",12),rep("cus",12)), other_seq_reads=c(mini_other_sequences$Reads,cus_other_seq$Reads))
# wilcoxon test requires exactly 2 levels of groups factors
mini_cus_other$db <- factor(mini_cus_other$db, levels=c("mini","cus"))

# wilcoxon test 
# alternative hypothesis, cus db with significantly lesser reads
# since tested on the same data set with different methods, paired in true
wilcox.test(mini_cus_other$other_seq_reads ~ mini_cus_other$db, paired=TRUE)
coin::wilcox_test(mini_cus_other$other_seq_reads ~ mini_cus_other$db, paired=TRUE) # not rejected
mini_cus_other %>% wilcox_effsize(other_seq_reads ~ db, paired = TRUE)

# std vs custom for other sequences reads
std_cus_other <- data.frame(db=c(rep("std",12),rep("cus",12)), other_seq_reads=c(other_sequences_std$Reads,cus_other_seq$Reads))
# wilcoxon test requires exactly 2 levels of groups factors
std_cus_other$db <- factor(std_cus_other$db, levels=c("std","cus"))

# wilcoxon test 
# alternative hypothesis, cus db with significantly lesser reads "alternative ="less"
# paired because tested the softwares on the same dataset
wilcox.test(std_cus_other$other_seq_reads ~ std_cus_other$db,paired=TRUE) # rejected
coin::wilcox_test(std_cus_other$other_seq_reads ~ std_cus_other$db,paired=TRUE)
std_cus_other
```
# microbial reads summary comparison
```{r}

# all three distribution for reads classified into the "other sequences" did not follow a normal distribution with the reject the null hypothesis at 0.05
shapiro.test(microbial_mini_summary$total_microbial)
shapiro.test(microbial_std_summary$total_microbial)
shapiro.test(microbial_cus_summary$total_microbial)

# use Wilcoxon test for difference with non-normal distributed groups

# mini vs custom for other sequences reads
mini_std_micro <- data.frame(db=c(rep("mini",12),rep("std",12)), microbial_reads=c(microbial_mini_summary$total_microbial,microbial_std_summary$total_microbial))
# wilcoxon test requires exactly 2 levels of groups factors
mini_std_micro$db <- factor(mini_std_micro$db, levels=c("mini","std"))

# wilcoxon test 
# alternative hypothesis, cus db with significantly lesser reads
# since tested on the same data set with different methods, paired in true
wilcox.test(mini_std_micro$microbial_reads ~ mini_std_micro$db,paired=TRUE,alternative = "less")
mini_std_micro

# mini vs custom for other sequences reads
mini_cus_micro <- data.frame(db=c(rep("mini",12),rep("cus",12)), microbial_reads=c(microbial_mini_summary$total_microbial,microbial_cus_summary$total_microbial))
# wilcoxon test requires exactly 2 levels of groups factors
mini_cus_micro$db <- factor(mini_cus_micro$db, levels=c("mini","cus"))

# wilcoxon test 
# alternative hypothesis, cus db with significantly lesser reads
# since tested on the same data set with different methods, paired in true
wilcox.test(mini_cus_micro$microbial_reads ~ mini_cus_micro$db, paired = TRUE,alternative="less")

mini_std_micro$microbial_reads
microbial_std_summary$total_microbial
```
# "unclassified" reads with mini database is significantly more than the mini and std databases reads
```{r}

# all three distribution for reads classified into the "other sequences" did not follow a normal distribution with the reject the null hypothesis at 0.05
shapiro.test(mini_unc$Reads)
shapiro.test(std_unc$Reads)
shapiro.test(cus_unc$Reads)

# use Wilcoxon test for difference with non-normal distributed groups

# mini vs custom for other sequences reads
mini_cus_unc <- data.frame(db=c(rep("mini",12),rep("cus",12)), unc_reads=c(mini_unc$Reads,cus_unc$Reads))
# wilcoxon test requires exactly 2 levels of groups factors
mini_cus_unc$db <- factor(mini_cus_unc$db, levels=c("mini","cus"))

# wilcoxon test 
# alternative hypothesis, cus db with significantly lesser reads
# since tested on the same data set with different methods, paired in true
wilcox.test(mini_cus_unc$unc_reads ~ mini_cus_unc$db, paired=TRUE, alternative="greater") # not rejected

# std vs custom for other sequences reads
mini_std_unc <- data.frame(db=c(rep("mini",12),rep("std",12)), unc_reads=c(mini_unc$Reads,std_unc$Reads))
# wilcoxon test requires exactly 2 levels of groups factors
mini_std_unc$db <- factor(mini_std_unc$db, levels=c("mini","std"))

# wilcoxon test 
# alternative hypothesis, cus db with significantly lesser reads
# paired because tested the softwares on the same dataset
wilcox.test(mini_std_unc$unc_reads ~ mini_std_unc$db, paired =TRUE, alternative="greater") # not rejected

mini_cus_unc
```

# Genus Level microbial composition standard db
```{r}

absolute_path = paste(path,"standard/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  genus_absolute <- current_sample %>% subset(V4 == "G") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  genus_absolute <- genus_absolute #%>% subset(c(trimed!="Homo")) %>% subset(V2 >= 100) # deselect genus belongs to Eukrayota
  colnames(genus_absolute) <- c("genus", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    genus_all_samples <-  genus_absolute

  } else {
    genus_all_samples <- full_join(genus_all_samples, genus_absolute,by="genus")

  }
  i = i + 1
}

# save(genus_all_samples, file ="standard_absolute_genus_micro_only.Rdata")

genus_all_samples[is.na(genus_all_samples)] <- 0
keys_to_gather <- colnames(genus_all_samples)[-1] # gather samples, exclude "genus" from colnames

kraken_std_genus <- genus_all_samples
# prepare dataframe for ggplot
genus_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

# genus I want to factored in order so they can be assigned with a specific color during plotting
all_genus <- unique(genus_all_samples$genus) # allgenus classified
order_genus <- c("Homo","Xanthomonas","Pseudomonas","Pasteurella","Streptomyces","Calothrix","Andhravirus","Mycoplasma","Spiroplasma","Bacillus","Staphylococcus","Lactobacillus","Clostridium","Mycobacterium","Bordetella","Leptospira","Yersinia","Escherichia","Bartonella","Bacteroides","Microcystis","Nostoc","Mycobacteroides","Alistipes","Parabacteroides","Anabaena","Dolichospermum","Faecalibacterium","Sphaerospermopsis","Prevotella","Achromobacter") 
remaining_genus <- all_genus %>% data.frame() %>% subset(!c(all_genus %in% order_genus)) # deselect genus subsetted to put in order
remaining_genus <- remaining_genus$. # data.frame to vector

order <- unique(genus_all_samples$samples)
genus_all_samples$samples <- factor(genus_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
genus_all_samples$genus <- factor(genus_all_samples$genus, level=c(order_genus,remaining_genus))

# number of reads classified into microbiome composition (exclude homo and rattus)
genus_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>%subset(genus != "Rattus" & genus !="Homo") %>% summarise(sum_reads=sum(Reads), sum_percentage=sum(Percentage)) -> std_micro_comp
std_micro_comp

```

# Genus Level microbial composition customized db
```{r}

absolute_path = paste(path,"custom/absolute/",sep="")
files <- list.files(absolute_path)


i=0

# iterate through all sample's result for Domain info
for (file in files){
  current_sample <- read.csv(paste(absolute_path,file,sep=""),sep="\t",header=FALSE)
  genus_absolute <- current_sample %>% subset(V4 == "G") %>% mutate(trimed = trimws(V6))  %>% select(c(trimed,V2))
  genus_absolute <- genus_absolute #%>% subset(trimed!="Rattus" & trimed !="Homo") #%>% subset(V2 >= 300) # deselect genus belongs to Eukrayota
  colnames(genus_absolute) <- c("genus", unlist(strsplit(file,".",fixed = TRUE))[1])
  
  if (i == 0){ # if is the first sample, make the table the aggregated data table
    genus_all_samples <-  genus_absolute

  } else {
    genus_all_samples <- full_join(genus_all_samples, genus_absolute,by="genus")

  }
  i = i + 1
}

# save(genus_all_samples, file ="standard_absolute_genus_micro_only.Rdata")

genus_all_samples[is.na(genus_all_samples)] <- 0
keys_to_gather <- colnames(genus_all_samples)[-1] # gather samples, exclude "genus" from colnames
kraken_cus_genus <- genus_all_samples
# prepare dataframe for ggplot
genus_all_samples %<>% gather(keys_to_gather, key="samples", value="Reads")

# genus I want to factored in order so they can be assigned with a specific color during plotting
all_genus <- unique(genus_all_samples$genus) # allgenus classified
order_genus <- c("Homo","Rattus","Xanthomonas","Pseudomonas","Pasteurella","Streptomyces","Calothrix","Andhravirus","Mycoplasma","Spiroplasma","Bacillus","Staphylococcus","Lactobacillus","Clostridium","Mycobacterium","Bordetella","Leptospira","Yersinia","Escherichia","Bartonella","Bacteroides","Microcystis","Nostoc","Mycobacteroides","Alistipes","Parabacteroides","Anabaena","Dolichospermum","Faecalibacterium","Sphaerospermopsis","Prevotella","Achromobacter") 
remaining_genus <- all_genus %>% data.frame() %>% subset(!c(all_genus %in% order_genus)) # deselect genus subsetted to put in order
remaining_genus <- remaining_genus$. # data.frame to vector

genus_all_samples$samples <- factor(genus_all_samples$samples, levels = c("R22_K","R26_K","R27_K","R28_K","R22_L","R26_L","R27_L","R28_L","R22_S","R26_S","R27_S","R28_S"))
genus_all_samples$genus <- factor(genus_all_samples$genus, level=c(order_genus,remaining_genus))

# subsetting taxonomy with more than 1% in relative composition from each sample 
genus_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total)# %>% subset(Percentage >= 0.010) ->major_genus

# number of reads classified into microbiome composition (exclude homo and rattus)
genus_all_samples %>% group_by(samples) %>%
  mutate(Total=sum(Reads), Percentage = Reads/Total) %>%subset(genus != "Rattus" & genus !="Homo") %>% summarise(sum_reads=sum(Reads), sum_percentage=sum(Percentage)) -> cus_micro_comp
cus_micro_comp
```

#combine standard and customized
```{r}
# kraken_std_genus
# write.csv(kraken_std_genus, "/Users/rachel/Desktop/kraken_std_genus.csv")
# kraken_cus_genus
# write.csv(kraken_cus_genus, "/Users/rachel/Desktop/kraken_cus_genus.csv")


```

```{r}

# all three distribution for reads classified into the "other sequences" did not follow a normal distribution with the reject the null hypothesis at 0.05
shapiro.test(cus_micro_comp$sum_reads)
shapiro.test(std_micro_comp$sum_reads) # normal not rejected


# use Wilcoxon test for difference with non-normal distributed groups

# mini vs custom for other sequences reads
std_cus_mic <- data.frame(db=c(rep("cus",12),rep("std",12)), mic_reads=c(cus_micro_comp$sum_reads,std_micro_comp$sum_reads))
# wilcoxon test requires exactly 2 levels of groups factors
std_cus_mic$db <- factor(std_cus_mic$db, levels=c("cus","std"))

# wilcoxon test 
# alternative hypothesis, cus db with significantly lesser reads
# paired because tested the softwares on the same dataset
wilcox.test(std_cus_mic$mic_reads ~ std_cus_mic$db, paired=TRUE) # not rejected


```